### ISSUE TITLE
Feature Request: Printer Emulation

### ISSUE THREAD

**Issue by DanTheMan827** [dosbox-staging/dosbox-staging#686]
Hal 9000's mega builds had printer emulation, it would be nice if this feature could be added along with potentially making some improvements like printing to PDF and asking where to save instead of that being set in the config file

**Comment by Burrito78** [comment]
eXoDOS and eXoWin3x use printer emulation too, so yay.

**Comment by johnnovak** [comment]
We need a volunteer! And yeah, it would be cool 😎

**Comment by Torinde** [comment]
DOSbox-X [supports printing](http://dosbox-x.com/wiki/Guide%3ASetting-up-printing-in-DOSBox%E2%80%90X) as well with further printer enhancements in [DOSBox-X-App](https://mendelson.org/dosboxapp.html) by @emendelson.

[vDos](https://github.com/joncampbell123/dosbox-x/discussions/5807) is also focused on printing support.

Related: exoscoriae/eXoDOS/issues/3368, [eXoDOS Staging project card](https://github.com/dosbox-staging/dosbox-staging/projects/7#card-87453225)

[h-a-l-9000](https://web.archive.org/web/20100219145734/http://home.arcor.de/h-a-l-9000/) ([branch in this repo](https://github.com/dosbox-staging/dosbox-staging/tree/forks/hal9k-megabuild6-r3661)) supports color printer - (one of) the last of its features not implemented in any contemporary DOSbox fork (the other feature being Native IPX DOSBox-to-real PC, which I'm not sure if X supports already).

**Comment by rderooy** [comment]
This may be of interest. Epson ESC/P and ESC/P2 to PDF. https://github.com/RWAP/PrinterToPDF

**Comment by johnnovak** [comment]
> This may be of interest. Epson ESC/P and ESC/P2 to PDF. https://github.com/RWAP/PrinterToPDF

Nice find @rderooy . But... depends on ImageMagick 😰

**Comment by MaddTheSane** [comment]
I know Boxer (Hi) has printer emulation. I don't know where it came from, but @alunbestor hooked it up to Boxer. You can see [my fork](https://github.com/MaddTheSane/dosbox-staging/tree/boxer-compat) for the changes he implemented.

Note that the last merge I did was in 2022… to a commit from 2021.

**Comment by Grounded0** [comment]
Does anyone have a list of DOS games with printing support?

**Comment by rderooy** [comment]
First priority would be to add support for simply writing the LPT output to a file, this would already allow simple text output to be written to a file on the host, as well as allowing DOS applications to write more complex ESC/P, PCL or PostScript output, for which you may have external converters on the host.

Assuming actual printer emulation, I think the following printer types would be good candidates:
- Generic Text printer
- IBM ProPrinter
- Epson ESC/P and ESC/P2
- PostScript
- HP PCL

Text, ESC/P, ESC/P2 and PostScript are supported by DOSBox-X and 86Box. IBM ProPrinter is partly supported by DOSBox-X as part of the Epson printer emulation.

**Comment by Grounded0** [comment]
Just capturing the raw output coming into the LPT would be fine. It doesn't need to be processed on our end in any way as it can be further processed on the host once it has the data.

**Comment by mbenkmann** [comment]
Rhere should be the option to specify a program that would be started the first time something is written to LPT and gets the LPT data written to its standard input. This could be done by specifying the path with a "|" in front like "| command --to --execute".

**Comment by rderooy** [comment]
One potential issue to be aware of is that the text output would default to the codepage set in Staging (or a guest OS), which is likely to be a source of problems. So you would probably want both a "raw text" output and a "utf-8 text" output. Although the later would not work with a guest OS as Staging would not have any way of knowing the codepage in that case.

**Comment by johnnovak** [comment]
Core members interested in printing support, or non-core members interested in printing support _and_ becoming core members, please step forward 😎 

Realistically, we need a new guy for this extremely niche feature. Our current team has plenty to do, are the issues we're tackling are far more important.

**Comment by Python-Exoproject** [comment]
Couple more print games: Newsroom, The (1985) and Newsroom Pro, The (1989)

**Comment by Atndlr** [comment]
Der Planer (1994) also utilizes printer for gameplay features.

**Comment by Grounded0** [comment]
Just noticed MobyGames has printer support attribute. This list is comprehensive with 242 entries for DOS (including Add-ons and NSFW games):

https://www.mobygames.com/game/attribute:114/platform:dos/sort:title/page:1/

**Comment by uturuncuayaku** [comment]
Hi! I'm interested in tackling this issue. I'm working through a structured program at CodePath and would like to take this on as my project for this semester as I finish my other online studies. 
Is this still something a soon to be CS graduate could tackle over the next four to eight weeks? Are you still interested in pursuing this enhancement?

It's not glorious or popular but I love learning new or different concepts especially printing because it could definitely be a cool way to interact with a retro game! I really like Grounded0's approach, capture it straight to LPT and then to a file. I'll be checking out the command `man cupsfilter`

**Comment by johnnovak** [comment]
@uturuncuayaku Thanks for your interest. Yes, this ticket is still waiting for someone interested in learning more about DOS printing and adding support for it.

I can't really say how long this will take for you — it depends on skill level, familiarity with the project, the scope of the initial implementation, etc. But I'd say you could get a lot done in two months.

The thing is, we haven't really scoped this out yet. There are lots of ideas and reference implementations in the above comments, but first of all, we'll need to define the scope and work from there. So the high-level plan would be:

- The primary goal is to add printing support to games that can print out character sheets, the command history in text adventures, and so on. The ideal outcome is if DOSBox Staging could just generate PDF files directly.
- Adding printing support for non-gaming applications is not a goal, or a secondary goal at best (we'll probably get some level of support for free after implementing support for the games).
- Look at [this list of games](https://docs.google.com/spreadsheets/d/1ekGudKxgvi6AmBIFu3wbpdfbL2irlwWepngAVi7-wqY/edit?gid=0#gid=0) that support printing. The goal here would be to set these games up, play around with them, and figure out how they support printing, then look at all the different ways games do printing.
- From this research, we'll see if we need to support more ways to print, or just one. If there's more, we can perhaps just implement the most common one first.
- After this, we can think about the technical implementation. The ideal outcome is if Staging generates PDF files after we instruct the game to print something. We could potentially use existing libraries out there that can help us, or check the existing reference implementations in other DOSBox variants.

So there's a fair bit of research involved in this. If you wanna get started on it, I recommend starting a new discussion ticket (https://github.com/dosbox-staging/dosbox-staging/discussions) and document the research there, then we can discuss.

**Comment by mbenkmann** [comment]
@uturuncuayaku As I suggested in my comment further above, I would implement this via a config option to specify an external program. When LPT is first written to, start that program and pipe everything into its stdin. This implementation should be easy to do within the scope of your project. Once that is done, the external program(s) can be developed separately from DOSBox which allows for more rapid development and support for a variety of output options, not just PDF. It also allows the creation of more specialized programs that could for instance emulate hardware other than printers that was connected to LPT.

**Comment by johnnovak** [comment]
> [@uturuncuayaku](https://github.com/uturuncuayaku) As I suggested in my comment further above, I would implement this via a config option to specify an external program. When LPT is first written to, start that program and pipe everything into its stdin. This implementation should be easy to do within the scope of your project. Once that is done, the external program(s) can be developed separately from DOSBox which allows for more rapid development and support for a variety of output options, not just PDF. It also allows the creation of more specialized programs that could for instance emulate hardware other than printers that was connected to LPT.

No, it cannot be an external program like that. The solution must be integrated into DOSBox Staging. Ease of use is important and this needs to work on Windows, too. Let's not confuse our volunteer please, I've laid out the plan already.

**Comment by mbenkmann** [comment]
@johnnovak Windows can pipe to external programs, too, and you ship a default preconfigured so the user never has to touch the config.

**Comment by uturuncuayaku** [comment]
I started a discussion, https://github.com/dosbox-staging/dosbox-staging/discussions/4931 
I agree with ease of use, I had an old printer with the cool strips on the left and right margins and having the emulator start printing mid-game would be pretty cool.

**Comment by johnnovak** [comment]
> [@johnnovak](https://github.com/johnnovak) Windows can pipe to external programs, too, and you ship a default preconfigured so the user never has to touch the config.

I don't want an external converter for this. I want to keep this simple and contained.

**Comment by mbenkmann** [comment]
Unfortunately printing didn't used to be simple under DOS. There was no standard. Every printer was different. And then there's other hardware like plotters. You just want to pick one printer with one character set (ASCII, of course, ignoring everything that's not the US) and implement that. Then you might just as well not support printing at all.

**Comment by johnnovak** [comment]
> Unfortunately printing didn't used to be simple under DOS. There was no standard. Every printer was different. And then there's other hardware like plotters. You just want to pick one printer with one character set (ASCII, of course, ignoring everything that's not the US) and implement that. Then you might just as well not support printing at all.

You don't understand. This is for supporting _games_ that have a printing feature. We don't care about the rest.

All games just send ASCII streams AFAIK.

Look at our goals and non-goals. We don't care about printing in appplications — just the ~30 games that support printing.

https://www.dosbox-staging.org/about/#goals

PS: I might have been wrong and some games might actually print graphics. But Boxer (an old abandoned DOSBox fork) has full built-in support for that. We want the same level of support — built into Staging: See [here](https://github.com/dosbox-staging/dosbox-staging/discussions/4931#discussioncomment-17243648).

**Comment by dux119-cmd** [comment]
> Then you might just as well not support printing at all.

That's rather fatalistic, no?

Most people printing from DOS (games, text editor, word processor) only had a dot matrix printer, and therefore dot matrix emulation is a whole lot better than nothing. 

I suppose fancy postscript laser printers could be emulated at some point, too - but in the early 90s they cost more than an entire DOS PC, and were most likely owned by professionals or businesses. $4000+ USD https://web.archive.org/web/20060910232053/http://www.printerworks.com/Catalogs/BX-Catalog/BX-NewUsers.html

**Comment by johnnovak** [comment]
> Most people printing from DOS (games, text editor, word processor) only had a dot matrix printer, and therefore dot matrix emulation is a whole lot better than nothing.

Exactly, dot matrix printer emulation is all what we need. See my post here:
https://github.com/dosbox-staging/dosbox-staging/discussions/4931#discussioncomment-17243648

Boxer had a cool implementation in 2012:
http://blog.boxerapp.com/2012/11/12/pc-load-letter/

<img width="494" height="494" alt="Image" src="https://github.com/user-attachments/assets/e9efd1fd-a338-4dfb-b7a0-882a99c39640" />

That would be a super cool and fun feature to emit PDF files like this! 😎 That's our gold-standard goal.

> I suppose fancy postscript laser printers could be emulated at some point, too - but in the early 90s they cost more than an entire DOS PC

Yeah, I'm iron-fisted about the scope. We're all about games. Everything else is _ruthlessly_ out of scope 😎 There are VMs for other use cases.

**Comment by mbenkmann** [comment]
> > Then you might just as well not support printing at all.
> 
> That's rather fatalistic, no?
> 

Right now there is no printing. There is one ticket. This one. It's been like this for years. It's, in a way, a clean solution. I believe if you add limited printer support, you will inevitably end up with a stream of tickets "Umlauts are garbled.", "Add support for Japanese printer XYZ", "Add plotter support".

From a technical standpoint I really do believe that no printer support is better than limited printer support that only creates desire for more.

**Comment by johnnovak** [comment]
@mbenkmann You've shared your opinion, everyone is welcome to do that. We've noted what you "believe". But what you're doing now is bordering on aggressive. The project leaders decide the direction, so let it go.

**Comment by Grounded0** [comment]
Personally, I would be just fine with LPT being dumped into a file. I could use CUPS to convert it to PDF. If there are no takers for the larger job, one could start with implementing this baseline since future improvements build upon this foundation.

Also, you could get PostScript laser printers with the genuine Adobe NPS (Native PostScript) engine starting from 1,000 USD in the early 1990s. My parents had one back then, so these things weren't luxuries if you printed a lot and needed impeccable output quality and high reliability.

PostScript and PDF share a similar imaging model, so conversion to PDF is fast and extremely accurate.

**Comment by uturuncuayaku** [comment]
That really seems like the best solution is to only go as far out as
printers from CUPS. I could keep the fork alive as well as being the one
stop shop for DOS game printing. I don't see an issue with that, if it's
possible I could try otherwise the default printing should be CUPS/PDF.
Targeting printers from that era would be fun. I'm sure there are plenty of
drivers around that if someone has a request that they submit the old
driver to be reverse engineered or least help them realize it's a hard ask
for someone without the data sheet of an ancient printer from another
country.

**Comment by rderooy** [comment]
> Personally, I would be just fine with LPT being dumped into a file. I could use CUPS to convert it to PDF. If there are no takers for the larger job, one could start with implementing this baseline since future improvements build upon this foundation.
> 
> Also, you could get PostScript laser printers with the genuine Adobe NPS (Native PostScript) engine starting from 1,000 USD in the early 1990s. My parents had one back then, so these things weren't luxuries if you printed a lot and needed impeccable output quality and high reliability.
> 
> PostScript and PDF share a similar imaging model, so conversion to PDF is fast and extremely accurate.

I think just starting with output to file would be a great first start. But one question immediately comes to mind, what about codepages? If no conversion is done, the output file will be in the codepage that is being emulated, which means that when you view the file on the host, you may get weird results. Automatic dosbox codepage to UTF-8 conversion (as long as you do not boot a guest OS) could perhaps be done, but you would have to ensure you only do it when the printer output is in text mode, and make sure to exclude any escape/control characters.

As to emulating different printers, 86Box has a very good range of options, including conversion to PDF for most.
- Generic Text Printer
- Generic ESC/P 2 Dot-Matrix Printer
- Generic PostScript Printer
- Generic PCL Printer

<img width="900" height="637" alt="Image" src="https://github.com/user-attachments/assets/ae527151-48a5-4b4b-b5b9-6f11d341e2e7" />
